flocker/node/agents/ebs.py

```diff
@@ -0,0 +1,198 @@
+        self.cluster_id = cluster_id
+
+    def _blockdevicevolume_from_ebs_volume(self, ebs_volume):
+        """
+        Helper function to convert Volume information from
+        EBS format to Flocker block device format.
+        """
+        return BlockDeviceVolume(
+            blockdevice_id=unicode(ebs_volume.id),
+            size=int(GB(ebs_volume.size).to_Byte().value),
+            host=None,
+            dataset_id=UUID(ebs_volume.tags[DATASET_ID_LABEL])
+        )
+
+    def _wait_for_volume(self, expected_volume,
+                         expected_status=u'available',
+                         time_limit=60):
+        """
+        Helper function to wait for up to 60s for given volume
+        to be in 'available' state.
+        """
+        start_time = time.time()
+        expected_volume.update()
+        while expected_volume.status != expected_status:
+            elapsed_time = time.time() - start_time
+            if elapsed_time < time_limit:
+                time.sleep(0.1)
+                expected_volume.update()
+            else:
+                raise Exception(
+                    'Timed out while waiting for volume. '
+                    'Expected Volume: {!r}, '
+                    'Expected Status: {!r}, '
+                    'Actual Status: {!r}, '
+                    'Elapsed Time: {!r}, '
+                    'Time Limit: {!r}.'.format(
+                        expected_volume, expected_status,
+                        expected_volume.status, elapsed_time,
+                        time_limit
+                    )
+                )
+
+    def _is_cluster_volume(self, cluster_id, ebs_volume):
+        """
+        Helper function to check if given volume belongs to
+        given cluster.
+        """
+        actual_cluster_id = ebs_volume.tags.get(CLUSTER_ID_LABEL)
+        if actual_cluster_id is not None:
+            actual_cluster_id = UUID(actual_cluster_id)
+            if actual_cluster_id == cluster_id:
+                return True
+        return False
+
+    def create_volume(self, dataset_id, size):
+        """
+        Create a volume on EBS. Store Flocker-specific
+        {metadata version, cluster id, dataset id} for the volume
+        as volume tag data.
+
+        :param String dataset_id: Dataset_id for the volume.
+        :param int size: Requested volume size in Bytes.
+
+        :return BlockDeviceVolume volume: Created volume in
+            BlockDeviceVolume format.
+        """
+        requested_volume = self.connection.create_volume(
+            size=int(Byte(size).to_GB().value), zone=self.zone)
+
+        # Stamp created volume with Flocker-specific tags.
+        metadata = {
+            METADATA_VERSION_LABEL: '1',
+            CLUSTER_ID_LABEL: unicode(self.cluster_id),
+            DATASET_ID_LABEL: unicode(dataset_id),
+        }
+        self.connection.create_tags([requested_volume.id],
+                                    metadata)
```

Comment:
Huh, so you have to tag the volume separately with EBS too.
Cinder allows metadata to be supplied when you create a volume, but I found it didn't work.
I wonder if we should think about what happens if the volume gets created but the `create_tag` API call fails.
Should we attempt to delete the volume? 
Or should we retry the tag operation?
With cinder you can choose the volume name which would allow you to identify volumes that had been only partially created, but boto doesn't seem to allow you to choose the name.
We should discuss with @exarkun and @itamarst